fix callback thread task lifetime

psychocoderHPC opened the pull request:
fix:  https://github.com/alpaka-group/alpaka/issues/2007 

Rewrite the callback thread to have a well-defined lifetime of the enqueued tasks and the moment when the future will be set after the tasks are executed.
The lifetime issue of data captured by a lambda introduced with https://github.com/alpaka-group/alpaka/pull/1870
is solved too and explicit destroying data is not required anymore.

- [x] depends on #2022

bernhardmgruber commented on include/alpaka/core/CallbackThread.hpp:
```diff
@@ -61,8 +61,9 @@ namespace alpaka::core
                 if(!m_thread.joinable())
                     startWorkerThread();
+                m_cond.notify_one();
             }
-            m_cond.notify_one();
+
```
Is that a necessary change? AFAIK notifying a condition variable outside the lock is actually the preferred implementation, since the notified thread can immediately acquire the lock.

psychocoderHPC replied:
Yes it is the same as in the de-structor we can miss the notification. The sanitizer issues got solved by this line.

bernhardmgruber replied:
Ah, I remember.  Alright then!

bernhardmgruber commented on test/unit/queue/src/QueueTest.cpp:
```diff
@@ -76,15 +76,18 @@ TEMPLATE_LIST_TEST_CASE(
     alpaka::enqueue(
         f.m_queue,
-        [&f, &callbackFinished]() noexcept
+        [&f, &callbackFinished, &inKernelIsNotEmptyBeforeSleep, &inKernelIsNotEmptyAfterSleep]() noexcept
```
```suggestion
        [&]() noexcept
```
I think I would just capture all now. Applies to the other lambdas as well.

psychocoderHPC replied:
I also thought about it but since we where explicit before I kept it.

psychocoderHPC replied:
done

j-stephan commented:
Still failing :-(

psychocoderHPC commented:
> Still failing :-(

Most likely the test is failing in `alpaka/test/unit/queue/src/CollectiveQueue.cpp`. I am currently checking whats going on there.
The test is providing a special queue DLR is using.

psychocoderHPC commented:
@bernhardmgruber @j-stephan 

I added a bunch of commits to this fix. The reason why I have not separated the fixes into different PR's is that all are related to the queue tests that we rely on a thread-safe implementation of the callback thread.

With all these fixes I do not see any more issues in the sanitizer.

psychocoderHPC commented:
Current CI issue:

```
38: enqueueBenchmark - alpaka::test::TestQueues - 1
38: -------------------------------------------------------------------------------
38: /home/runner/work/alpaka/alpaka/test/unit/queue/src/QueueTest.cpp:198
38: ...............................................................................
38: 
38: terminating with uncaught exception of type std::__1::future_error: The state of the promise has already been set.
38: benchmark name                       samples       iterations    estimated
38:                                      mean          low mean      high mean
38:                                      std dev       low std dev   high std dev
```

psychocoderHPC commented:
Successfully passed all tests once, I will force push again to check if it is a false positive.

bernhardmgruber commented:
This PR is starting to touch too many things. And I also don't believe all changes here are necessary, but it's harder to argue the bigger the PR gets. Can you please split off:
* the changes avoiding multithreaded use of Catch2
* the bugfix for resetting the event shared pointers while mutexes are being held.

Thx a lot!

psychocoderHPC commented:
> This PR is starting to touch too many things. And I also don't believe all changes here are necessary, but it's harder to argue the bigger the PR gets. Can you please split off:
> 
> * the changes avoiding multithreaded use of Catch2
> * the bugfix for resetting the event shared pointers while mutexes are being held.
> 
> Thx a lot!

I created #2023 and #2022 to pull out two fixes in the tests.
Changes in `CallbackThread.hpp` should not be split, I will reduce the number of commits with the rebase after both dependency PRs are merged.

psychocoderHPC committed:
fix callback thread task lifetime

fix:  #2007 #1936

Rewrite the callback thread to have a well defined lifetime of the
enqueued tasks and the moment when the future will be set after the
tasks is executed.
The lifetime issue of data captured by an lambda introduced with #1870
is solved to and explicit destroying data is not required anymore.

bernhardmgruber commented on include/alpaka/core/CallbackThread.hpp:
```diff
@@ -38,11 +40,8 @@ namespace alpaka::core
-        // destroyed. Therefore, ensure that a submitted task does not extend the lifetime of any object that
-        // (transitively) holds the returned future. E.g. don't capture a shared_ptr to an alpaka object that stores
-        // the returned future. This is a cyclic dependency and creates a leak.
+        //! It is guaranteed that the task is fully destroyed before future result is set.
```
```suggestion
        //! It is guaranteed that the task is fully destroyed before the future's result is set.
```

bernhardmgruber commented on include/alpaka/core/CallbackThread.hpp:
```diff
@@ -16,7 +16,9 @@ namespace alpaka::core
 {
     class CallbackThread
     {
+        // std::packaged_task is used because std::function requires that the wrapped callable is copyable.
```
You can add a comment to replace it by `std::move_only_function` in C++23 :)

bernhardmgruber commented on include/alpaka/core/CallbackThread.hpp:
```diff
@@ -54,30 +53,35 @@ namespace alpaka::core
         auto submit(Task task) -> std::future<void>
         {
-            auto f = task.get_future();
+            auto tp = std::make_pair(std::move(task), std::promise<void>{});
```
Please add a comment here that we do NOT use the `packaged_task`'s future, because it keeps the task alive.

bernhardmgruber commented on include/alpaka/event/EventGenericThreads.hpp:
```diff
@@ -147,15 +147,16 @@ namespace alpaka
-
-                        // Nothing to do if it has been re-enqueued to a later position in the queue.
-                        if(enqueueCount == spEventImpl->m_enqueueCount)
+                        // The lock is released before the shared pointer is reset.
```
We don't need that comment anymore, do we?

bernhardmgruber commented on include/alpaka/event/EventGenericThreads.hpp:
```diff
@@ -302,10 +303,11 @@ namespace alpaka
-                            spEventImpl->wait(enqueueCount, lk2);
-                            spEventImpl.reset(); // avoid keeping the event alive as part of the background thread
-                                                 // task's future
+                            {
+                                // The lock is released before the shared pointer is reset.
+                                std::unique_lock<std::mutex> lk2(spEventImpl->m_mutex);
+                                spEventImpl->wait(enqueueCount, lk2);
+                            }
```
```suggestion
                                std::unique_lock<std::mutex> lk2(spEventImpl->m_mutex);
                                spEventImpl->wait(enqueueCount, lk2);
```

bernhardmgruber commented on include/alpaka/queue/cuda_hip/QueueUniformCudaHipRt.hpp:
```diff
@@ -203,12 +203,7 @@ namespace alpaka
-                        data->t();
-                        data.reset(); // destroy the task
-                    });
+                auto f = queue.m_callbackThread.submit([data = std::move(data)]() mutable { data->t(); });
```
```suggestion
                auto f = queue.m_callbackThread.submit([data = std::move(data)] { data->t(); });
```

bernhardmgruber commented on include/alpaka/test/event/EventHostManualTrigger.hpp:
```diff
@@ -180,13 +180,14 @@ namespace alpaka::trait
-                        { return (enqueueCount != spEventImpl->m_enqueueCount) || spEventImpl->m_bIsReady; });
-                    spEventImpl
-                        .reset(); // avoid keeping the event alive as part of the background thread task's future
+                    {
+                        // The lock is released before the shared pointer is reset.
+                        std::unique_lock<std::mutex> lk2(spEventImpl->m_mutex);
+                        spEventImpl->m_conditionVariable.wait(
+                            lk2,
+                            [spEventImpl, enqueueCount]
+                            { return (enqueueCount != spEventImpl->m_enqueueCount) || spEventImpl->m_bIsReady; });
+                    }
```
```suggestion
                        std::unique_lock<std::mutex> lk2(spEventImpl->m_mutex);
                        spEventImpl->m_conditionVariable.wait(
                            lk2,
                            [spEventImpl, enqueueCount]
                            { return (enqueueCount != spEventImpl->m_enqueueCount) || spEventImpl->m_bIsReady; });
```

psychocoderHPC commented:
@bernhardmgruber Thanks for the review, I change the parts you annotated and a few left over comments.
